packages/expo-apple-authentication/ios/EXAppleAuthentication/EXAppleAuthentication.m

```diff
@@ -0,0 +1,200 @@
+{
+  if (_moduleRegistry) {
+    [self invalidate];
+  }
+  _moduleRegistry = moduleRegistry;
+  _eventEmitter = [moduleRegistry getModuleImplementingProtocol:@protocol(UMEventEmitterService)];
+}
+
+- (NSArray<NSString *> *)supportedEvents
+{
+  return @[@"Expo.appleIdCredentialRevoked"];
+}
+
+- (void)invalidate
+{
+  _eventEmitter = nil;
+}
+
++ (BOOL)requiresMainQueueSetup
+{
+  return NO;
+}
+
+- (void)startObserving
+{
+  if (@available(iOS 13.0, *)) {
+    _hasListeners = YES;
+    [[NSNotificationCenter defaultCenter] addObserver:self
+                                             selector:@selector(credentialRevoked:)
+                                                 name:ASAuthorizationAppleIDProviderCredentialRevokedNotification
+                                               object:nil];
+  }
+}
+
+- (void)stopObserving
+{
+  if (@available(iOS 13.0, *)) {
+    _hasListeners = NO;
+    [[NSNotificationCenter defaultCenter] removeObserver:self
+                                                    name:ASAuthorizationAppleIDProviderCredentialRevokedNotification
+                                                  object:nil];
+  }
+}
+
+- (void)credentialRevoked:(NSNotification *)notification
+{
+  if (!_hasListeners) {
+    return;
+  }
+  [_eventEmitter sendEventWithName:@"Expo.appleIdCredentialRevoked" body:@{@"type": @"revoke"}];
+}
+
+UM_EXPORT_METHOD_AS(isAvailableAsync,
+                    isAvailableAsync:(UMPromiseResolveBlock)resolve
+                            rejecter:(UMPromiseRejectBlock)reject)
+{
+  if (@available(iOS 13.0, *)) {
+    return resolve(@(YES));
+  }
+  resolve(@(NO));
+}
+
+UM_EXPORT_METHOD_AS(requestAsync,
+                    requestAsync:(NSDictionary *)options
+                        resolver:(UMPromiseResolveBlock)resolve
+                        rejecter:(UMPromiseRejectBlock)reject)
+{
+  if (@available(iOS 13.0, *)) {
+    _promiseResolve = resolve;
+    _promiseReject = reject;
+    
+    ASAuthorizationAppleIDProvider* appleIDProvider = [[ASAuthorizationAppleIDProvider alloc] init];
+    ASAuthorizationAppleIDRequest* request = [appleIDProvider createRequest];
+
+    NSArray<NSNumber *> *requestedScopes = options[@"requestedScopes"];
+    NSNumber *requestedOperation = options[@"requestedOperation"];
+
+    request.requestedScopes = [EXAppleAuthenticationMappings importScopes:requestedScopes];
+    request.requestedOperation = [EXAppleAuthenticationMappings importOperation:requestedOperation];
+
+    if (options[@"user"]) {
+      request.user = options[@"user"];
+    }
+    if (options[@"state"]) {
+      request.state = options[@"state"];
+    }
+    
+    ASAuthorizationController* ctrl = [[ASAuthorizationController alloc] initWithAuthorizationRequests:@[request]];
+    ctrl.presentationContextProvider = self;
+    ctrl.delegate = self;
+    [ctrl performRequests];
+  } else {
+    reject(@"ERR_APPLE_AUTHENTICATION_UNAVAILABLE", @"This feature is not available on your iPhone.", nil);
```

Comment:
This error message won't make sense on an iPad nor or macOS, also "This feature" is imprecise and unhelpful in error logs (what feature is there a problem with?), and this error message is not meant to be user-facing:

```suggestion
    reject(@"ERR_APPLE_AUTHENTICATION_UNAVAILABLE", @"Apple authentication is not supported on this device.", nil);
```

Reply:
This still needs to be resolved (or maybe the commit hasn't been pushed yet)